lib/Dialect/Ibis/Transforms/IbisTunneling.cpp

```diff
@@ -0,0 +1,365 @@
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+
+#include "circt/Support/InstanceGraph.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace ibis;
+using namespace igraph;
+
+namespace {
+
+class TunnelingConversionPattern : public OpConversionPattern<PathOp> {
+public:
+  TunnelingConversionPattern(MLIRContext *context, igraph::InstanceGraph &ig)
+      : OpConversionPattern<PathOp>(context), ig(ig) {}
+
+  LogicalResult
+  matchAndRewrite(PathOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    return Tunneler(op, rewriter, ig).go();
+  }
+
+  struct Tunneler {
+    PathOp op;
+    ConversionPatternRewriter &rewriter;
+    igraph::InstanceGraph &ig;
+    mlir::StringAttr pathName;
+
+    // The PortInfo struct is used to keep track of the get_port ops that
+    // specify which ports needs to be tunneled through the hierarchy.
+    struct PortInfo {
+      // Name used for portrefs of this get_port in the instance hierarchy.
+      mlir::StringAttr portName;
+
+      // Source get_port op.
+      GetPortOp getPortOp;
+
+      PortRefType getType() {
+        return getPortOp.getPort().getType().cast<PortRefType>();
+      }
+      Type getInnerType() { return getType().getPortType(); }
+      ibis::Direction requestedDirection() { return getType().getDirection(); }
+    };
+
+    // MapVector to ensure determinism.
+    llvm::MapVector<StringAttr, PortInfo> portInfos;
+
+    // "Target" refers to the last step in the path which is the scoperef that
+    // all port requests are tunneling towards.
+    PathStepAttr target;
+    FlatSymbolRefAttr targetName;
+
+    Tunneler(PathOp op, ConversionPatternRewriter &rewriter,
+             igraph::InstanceGraph &ig)
+        : op(op), rewriter(rewriter), ig(ig) {}
+
+    // A mapping between requested port names from a ScopeRef and the actual
+    // portref SSA values that are used to replace the get_port ops.
+    // MapVector to ensure determinism.
+    using PortRefMapping = llvm::MapVector<StringAttr, Value>;
+
+    LogicalResult go() {
+      // Gather the required port accesses of the ScopeRef.
+      for (auto *user : op.getResult().getUsers()) {
+        auto getPortOp = dyn_cast<GetPortOp>(user);
+        if (!getPortOp)
+          return user->emitOpError() << "unknown user of a PathOp result - "
+                                        "tunneling only supports ibis.get_port";
+        portInfos[getPortOp.getPortSymbolAttr().getAttr()].getPortOp =
+            getPortOp;
+      }
+
+      InstanceGraphNode *currentContainer =
+          ig.lookup(cast<ModuleOpInterface>(op.getOperation()->getParentOp()));
+
+      llvm::SmallVector<PathStepAttr> path;
+      llvm::copy(op.getPathAsRange(), std::back_inserter(path));
+      genPortNames();
+      target = path.back();
```

Comment:
These look like class setup things which I think should be done in the constructor.